[RUNTIME][TRACE] Support trace primitive.

denis0x0D opened the pull request:
The patch is related to issue:
https://discuss.tvm.ai/t/idea-trace-expression/945

This patch allows to trace Tensor data at the runtime. 

Use case example:
**1. Write model.**
...
tvm::Array<tvm::Expr> shape = {n, n, n};
tvm::Tensor  C = tvm::compute (shape, [&] (tvm::Var i, tvm::Var j, tvm::Var k) {
 return tvm::trace ("A", {i,j, k}, A[i][j][k]) + B[i][j][k];
}
...

**2. Run model with specif data.**
output is
...
A [1][4][5] =  $actual_tensor_value
...

How it works:
tvm::trace adds ir::Call expr with type External and then lowers by llvm codegen part
to external call,  so, the runtime support is added. The runtime just processes added function expanding arguments by "var_arg" and writes to file specific data, by default it is a stdout, but log file could be specified by env variable TVM_TRACE, also tracing process could be disable/enable.
Example:
export TVM_TRACE=trace_log_file=path/to/file:process_tracing=1

tqchen commented:
Thanks for the contribution. We can likely further simplify this by using tvm's runtime system for callback. TVM support generated code to callback any PackedFunc (including ones from python).

Here is an rough example:
```python

import tvm
import numpy as np


@tvm.register_func
def my_debug(x, i):
    print("array=", x.asnumpy())
    return 0


x = tvm.placeholder((4,), name="x", dtype="int32")
xbuffer = tvm.decl_buffer(x.shape, dtype=x.dtype)

y = tvm.compute(x.shape, lambda i: tvm.call_packed("my_debug", xbuffer))
s = tvm.create_schedule(y.op)

print(tvm.lower(s, [x, y], binds={x:xbuffer}, simple_mode=True))

f = tvm.build(s, [xbuffer, y], binds={x:xbuffer})
xnd = tvm.nd.array(np.ones((4,), dtype=x.dtype))
ynd = tvm.nd.array(np.zeros((4,), dtype=y.dtype))

f(xnd, ynd)
```
In the above case, the call_packed calls my_debug in every iteration passing the array handle(xbuffer). Some further wrapping could make it looks nicer. The good part is, we can use any piece of python code to monitor the content of x
See more at https://docs.tvm.ai/dev/runtime.html

xqdan commented:
oh. we do have dump ir.
you can try like this.

with tvm.build_config(dump_pass_ir=True)
            run()

xqdan commented:
oh. we do have dump ir.
you can try like this.

with tvm.build_config(dump_pass_ir=True)
            run()

tqchen commented:
This is a good question :) 

The current principle is: if some API is stabilized and needed across languages, we bring it to C++. If something is in flux and not yet critical to the developments, we will keep it in the language bindings(e.g. python), let it plays out and wait until it stablizes.

Admittedly there are tradeoffs, having everything in c++ makes the API available to all languages, but it is also not necessarily flexible. For example, in the case of monitoring, printing things out to a log may not be as flexible as defining a customized callback and let the user do what they want to do. 

Having some of the things in python(for example, most of our test cases are) gives some speed up to the development process and allows us to write more test cases easily.

In both cases, we should document it clearly, so it would be really nice to have a tutorial to teach users on all the possible ways to do tracing

tqchen commented:
As a side note, in this particular case, the my_debug function can be registered in c++ as well, because the runtime system allows interop between languages

denis0x0D commented:
@tqchen thanks for the answer,
but anyway the tracing is require to write own runtime support, what do you think about generic solution written as shared library, with some options which could be applied by environment library. Looks like it would be an easy way to use it for the users.

For example "sanitizers" from llvm/compiler-rt have two parts:
1. Instrumentation pass.
2. Runtime support, which could disable/enable some features by environment variable.

The library can be build as unique dso and would require to manually link model with it.

tqchen commented:
The current features could likely be implemented using packed function calls, to define a set of tracing functions like you do but declare them as packed function.  Then optionally we could register these runtime function as a packed function, or registering different variant of it depending on whether tracing is on/off

So to summarize, I agree that tracing would need runtime support, but these runtime feature could likely be built around tvm packed function system

denis0x0D commented:
@tqchen got it now, thanks,
I'll update the patch.

xqdan commented:
@denis0x0D I'm considering move the dump ir to C++ side.

denis0x0D commented:
@tqchen  the strange thing I've faced with packed function - when I increase the shape of the buffer or so on.
For example:
```
import tvm
import numpy as np

@tvm.register_func
def my_debug(x):
    return 0

x = tvm.placeholder((1024, 1024), name="x", dtype="int32")
xbuffer = tvm.decl_buffer(x.shape, dtype=x.dtype)
y = tvm.compute(x.shape, lambda i, j: tvm.call_packed("my_debug", xbuffer))
s = tvm.create_schedule(y.op)

f = tvm.build(s, [xbuffer, y], binds={x:xbuffer})
xnd = tvm.nd.array(np.ones((1024, 1024), dtype=x.dtype))
ynd = tvm.nd.array(np.zeros((1024, 1024), dtype=y.dtype))
f(xnd, ynd)
```

I got Segmentation fault on 3 different platform and spent some time to debug, the segfault happens in hash table which registry the packed function. May be  I miss something or may it is a bug, it would be good to get confirmation from someone else. Also I have a backtrace of error and tests case on c++.

denis0x0D commented:
@tqchen  patch is updated, tests added, python api is added alongside with c++ api.

tqchen commented:
Thanks for the updates, the new set of changes makes use the packed_func runtime and reduces the runtime greatly. I am thinking how can even simplify this further, to consider most user's use case. Most of my reasoning comes from the fact that this is still a new feature and we may not settle down on what API user might prefer, so maybe it is a good idea to start quick mocks in python side. Here are some specific comments:

- I do feel the current flag parsing and the pre-defined prints are a bit redundant
   - If in most cases users could use the python runtime to debug, would it sufficient to provide a simple python callback function
   - Alternatively, a C++ PackedFunc version of log_to_file callback would be cleaner than the flag parsing + global print logic.
- For example, would it be sufficient to allow a user to register tvm_trace_callback into the PackedFunc, and only trigger the callback when necessary? 

Here is a mock up version of the code.

```python
from tvm.contrib import trace

@tvm.register_func("tvm.trace_callback", trace.log_to_file("mytrace.log")

# run the trace code
```

denis0x0D commented:
@tqchen thanks for review, I'll update it.

denis0x0D commented:
@tqchen thanks for review.

> Is this additional level of redirection necessary in the tvm_trace_runtime? Currently, the feature being implemented seems to be fine to just call call_packed.

As far as I understood, to be able to handle packed function, and take action only on those functions which are registered in global registry, we need an additional runtime machinery, in my case I've added global function "tvm.trace_runtime" which checks the global registry for the registered trace callbacks and calls it if necessary. In other case if we use call_packed inside computation node and the packed function is not registered we will get an error.
So, if we use only call_packed without wrapping, we should explicitly pass the name of runtime handler as first arg
`call_packed ("tvm.trace_runtime", "tvm_global_callback", expr)`
instead
`tvm.trace("tvm_global_callback", expr)`
Can you please clarify, what do you think, should I delete those wrapping ?

> One of the properties of trace is that we might want to return the same specific value x[i] when we trace the content, which might need some additional work in the compiler when seeing trace, this property seems not being reflected here.

Thanks for mention this, I completely forget about return value. Before updating the patch, I want make sure, that I get your right about "work in the compiler".
Seems to me, I just can add return value to TVMRetValue
`*ret = args[1];`
and this requires some cosmetic changes in code which lowers tvm_call_packed
https://github.com/dmlc/tvm/blob/master/src/pass/lower_tvm_builtin.cc#L254
From:
```
return Call::make(
        Int(32), intrinsic::tvm_call_packed_lowered,
        packed_args, Call::Intrinsic);
```
To:
```
return Call::make(
        op->type, intrinsic::tvm_call_packed_lowered,
        packed_args, Call::Intrinsic);

```
and to the call_packed python wrapper to create a Call node with the actual type of processed Expr.
Can you please confirm it, or there are could be some problems which I don't see, because a lot of logic depends on packed functions.
cc @grwlf 
Thanks.

denis0x0D commented:
@grwlf @tqchen updated.

yzhliu commented on src/runtime/trace.cc:
```diff
@@ -0,0 +1,37 @@
+          (*pf)(args[1], args[2], args[3], args[4]);
+        } else if (args.size() == 6) {
+          (*pf)(args[1], args[2], args[3], args[4], args[5]);
+        }
```
else throw error?

denis0x0D replied:
@yzhliu thanks for review, moved the check for the arguments count to the line 16, before the call.

yzhliu commented on python/tvm/intrin.py:
```diff
@@ -488,6 +488,43 @@ def _rule_float_direct(op):
     return None
 
+@_register_func("tvm.default_trace_action")
+def tvm_default_trace_action(*args):
```
suggest to name it `_tvm_default_trace_action`

denis0x0D replied:
thanks for review, updated.

denis0x0D commented:
@yzhliu @grwlf updated.

tqchen commented:
Let us double check the implementation of call_packed_lowered to make sure that the return value of the trace function is lowered correctly

denis0x0D commented:
@tqchen should I add more tests ?

tqchen commented:
@denis0x0D never mind, I checked the current code and it should support return value correctly.

denis0x0D commented:
@tqchen thanks for review, I agree, your strategy is faster, without additional check in global registry. 
I'll update the patch.

denis0x0D commented:
@tqchen updated regarding the latest comment, the tracing value explicitly sets into return value in codegen stage for trace intrinsic.
cc @yzhliu @grwlf

tqchen commented:
@yzhliu @grwlf Please take another round of look

grwlf commented:
Looks good to me. My use-cases all work.

tqchen commented:
@denis0x0D thanks for the change. The logic can be improved further, here are some parts that can be improved

- The condition handling is done inside the CallPacked and is a bit obscured because the behavior depends on op's name.
- The returning of the traced value is done via load from stored argument buffer, while we could have directly done so in intrinsic handling.

Here is one possible way to improve the current logic.
- When lowering trace, keep the traced value as the last argument.
- Pass an optional return value as an argument of CallPacked, if that is not null, make sure CallPacked return the corresponding value

denis0x0D commented:
@tqchen thanks for review, sorry for the late response, I was on a holidays.
I've updated the patch regarding to your comments:
1. Separated the traced machinery from tvm_packed.
2. The traced value passes as the last argument when lowering.
3. Added additional conditional branch for return value in codegen. In case the TVMRetValue is zero we return traced value, otherwise we return the value specified by the user. In this case it's easy for user to redefine the traced value from python code:
```
@tvm.register_func("call_back")
def call_back(x):
   print (x)
   return user_defined_value
```
Also tests added for that functional too.
I will be appreciate for any feedback.
Thanks.
cc @grwlf @yzhliu

denis0x0D commented on src/codegen/llvm/codegen_cpu.cc:
```diff
@@ -562,6 +562,72 @@ llvm::Value* CodeGenCPU::CreateCallPacked(const Call* op) {
+  llvm::Value *cmp;
+  llvm::Type *rvalue_type = rvalue->getType();
+  // The cmp instruction depends on processing type.
+  if (rvalue_type->isFloatTy() || rvalue_type->isDoubleTy() ||
```
thanks for mention tcode, I was spending couple of hours to understand how to check return value, and didn't find a better solution to return 0 from call back when we want to process origin value, otherwise return user-defined value, but the true way to do it is by checking tcode.

denis0x0D commented on src/codegen/llvm/codegen_cpu.cc:
```diff
@@ -562,6 +562,72 @@ llvm::Value* CodeGenCPU::CreateCallPacked(const Call* op) {
+llvm::Value *CodeGenCPU::CreateCallTracePacked(const Call *op) {
+  using llvm::BasicBlock;
+  CHECK_EQ(op->args.size(), 6U);
+  std::string func_name = op->args[0].as<StringImm>()->value;
```
Thanks, done. Also added some additional arguments as rvalue, r_type and function success returns basic block, so it's easy to create a phi node in this case.

denis0x0D commented:
@tqchen thanks for review, I've updated the patch regarding to latest comments.

tqchen commented:
Thanks @denis0x0D @grwlf @yzhliu !This is now merged !
